File: analysis.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

def main():
    print("========================================")
    print("      MOVIE RATING ANALYSIS STARTED     ")
    print("========================================\n")
    
    # 1. Load Data
    print("[1] Loading datasets...")
    try:
        ratings = pd.read_csv('ratings.csv')
        movies = pd.read_csv('movies.csv')
    except FileNotFoundError as e:
        print(f"Error: {e}")
        print("Please make sure 'ratings.csv' and 'movies.csv' are in the project folder.")
        return

    # 2. Preprocess Data
    print("[2] Preprocessing data...")
    ratings.dropna(inplace=True)
    movies.dropna(inplace=True)
    # Merge datasets on 'movieId'
    movie_data = pd.merge(ratings, movies, on='movieId')

    # Create Features (Popularity and Average Rating)
    movie_stats = movie_data.groupby('title').agg({'rating': ['count', 'mean']})
    movie_stats.columns = ['rating_count', 'average_rating']
    movie_stats.reset_index(inplace=True)

    # Setup output directories
    if not os.path.exists('plots'):
        os.makedirs('plots')
        print("    -> Created 'plots' folder for saving visualizations.")

    # 3. Exploratory Data Analysis & Visualizations
    print("[3] Generating visualizations and exporting results...\n")

    # A. Rating Distribution
    print("    -> Plotting: Rating Distribution...")
    plt.figure(figsize=(8, 5))
    sns.histplot(movie_data['rating'], bins=10, kde=True, color='skyblue')
    plt.title('Distribution of Movie Ratings')
    plt.xlabel('Rating (0.5 to 5.0)')
    plt.ylabel('Count')
    plt.tight_layout()
    plt.savefig('plots/1_rating_distribution.png')
    plt.close()

    # B. Top Rated Movies (Min 50 Ratings)
    print("    -> Extracting: Top Rated Movies...")
    top_movies = movie_stats[movie_stats['rating_count'] >= 50].sort_values(by='average_rating', ascending=False)
    
    print("\n--- Top 10 Movies (Minimum 50 Ratings) ---")
    print(top_movies[['title', 'rating_count', 'average_rating']].head(10).to_string(index=False))
    print("------------------------------------------\n")
    
    # Exporting the full sorted list to CSV
    top_movies.to_csv('top_rated_movies_report.csv', index=False)
    print("    -> Saved 'top_rated_movies_report.csv'.")

    # C. Popularity vs Rating
    print("    -> Plotting: Popularity vs Average Rating...")
    plt.figure(figsize=(8, 5))
    sns.scatterplot(x='rating_count', y='average_rating', data=movie_stats, alpha=0.5, color='coral')
    plt.title('Popularity vs Average Rating')
    plt.xlabel('Number of Ratings (Popularity)')
    plt.ylabel('Average Rating')
    plt.tight_layout()
    plt.savefig('plots/2_popularity_vs_rating.png')
    plt.close()

    # D. Genre Analysis
    print("    -> Plotting: Genre Analysis...")
    # Splitting the pipe separated genre strings efficiently
    movie_genres = movie_data.assign(genres=movie_data['genres'].str.split('|')).explode('genres')
    genre_stats = movie_genres.groupby('genres').agg({'rating': ['count', 'mean']})
    genre_stats.columns = ['rating_count', 'average_rating']
    # Filtering genres with substantial representation
    genre_stats = genre_stats[genre_stats['rating_count'] > 1000].sort_values(by='average_rating', ascending=False)

    plt.figure(figsize=(12, 6))
    sns.barplot(x=genre_stats.index, y='average_rating', data=genre_stats, hue=genre_stats.index, palette='viridis', legend=False)
    plt.xticks(rotation=45)
    plt.title('Average Rating by Genre (Min 1000 Ratings)')
    plt.ylabel('Average Rating')
    plt.tight_layout()
    plt.savefig('plots/3_genre_analysis.png')
    plt.close()

    # E. User Behavior Analysis
    print("    -> Plotting: User Behavior Analysis...")
    user_stats = movie_data.groupby('userId').agg({'rating': ['count', 'mean']})
    user_stats.columns = ['rating_count', 'average_rating']

    plt.figure(figsize=(8, 5))
    sns.histplot(user_stats['average_rating'], bins=20, kde=True, color='purple')
    plt.title('Distribution of User Average Ratings')
    plt.xlabel('Average Rating given by User')
    plt.ylabel('Count of Users')
    plt.tight_layout()
    plt.savefig('plots/4_user_behavior.png')
    plt.close()

    print("\n========================================")
    print("          ANALYSIS COMPLETELY FINISHED        ")
    print("========================================")
    print("All visualizations have been cleanly saved in the 'plots/' directory.")
    print("Ready for your presentation/review!")

if __name__ == "__main__":
    main()

File: analysis.R
# Load required libraries
if(!require(dplyr)) install.packages("dplyr", repos = "http://cran.us.r-project.org")
if(!require(ggplot2)) install.packages("ggplot2", repos = "http://cran.us.r-project.org")
if(!require(readr)) install.packages("readr", repos = "http://cran.us.r-project.org")
if(!require(stringr)) install.packages("stringr", repos = "http://cran.us.r-project.org")
if(!require(tidyr)) install.packages("tidyr", repos = "http://cran.us.r-project.org")

library(dplyr)
library(ggplot2)
library(readr)
library(stringr)
library(tidyr)

main <- function() {
  cat("========================================\n")
  cat("      MOVIE RATING ANALYSIS STARTED     \n")
  cat("========================================\n\n")
  
  # 1. Load Data
  cat("[1] Loading datasets...\n")
  if (!file.exists('ratings.csv') || !file.exists('movies.csv')) {
    stop("Error: Please make sure 'ratings.csv' and 'movies.csv' are in the project folder.")
  }
  
  ratings <- read_csv('ratings.csv', show_col_types = FALSE)
  movies <- read_csv('movies.csv', show_col_types = FALSE)
  
  # 2. Preprocess Data
  cat("[2] Preprocessing data...\n")
  ratings <- drop_na(ratings)
  movies <- drop_na(movies)
  
  # Merge datasets on 'movieId'
  movie_data <- inner_join(ratings, movies, by = 'movieId')
  
  # Create Features (Popularity and Average Rating)
  movie_stats <- movie_data %>%
    group_by(title) %>%
    summarise(
      rating_count = n(),
      average_rating = mean(rating)
    ) %>%
    ungroup()
  
  # Setup output directories
  if (!dir.exists('plots')) {
    dir.create('plots')
    cat("    -> Created 'plots' folder for saving visualizations.\n")
  }
  
  # 3. Exploratory Data Analysis & Visualizations
  cat("[3] Generating visualizations and exporting results...\n\n")
  
  # A. Rating Distribution
  cat("    -> Plotting: Rating Distribution...\n")
  p1 <- ggplot(movie_data, aes(x = rating)) +
    geom_histogram(bins = 10, fill = 'skyblue', color = 'black', alpha = 0.8) +
    labs(title = 'Distribution of Movie Ratings', x = 'Rating (0.5 to 5.0)', y = 'Count') +
    theme_minimal()
  ggsave('plots/1_rating_distribution.png', plot = p1, width = 8, height = 5)
  
  # B. Top Rated Movies (Min 50 Ratings)
  cat("    -> Extracting: Top Rated Movies...\n")
  top_movies <- movie_stats %>%
    filter(rating_count >= 50) %>%
    arrange(desc(average_rating))
  
  cat("\n--- Top 10 Movies (Minimum 50 Ratings) ---\n")
  print(head(top_movies, 10))
  cat("------------------------------------------\n\n")
  
  # Exporting the full sorted list to CSV
  write_csv(top_movies, 'top_rated_movies_report.csv')
  cat("    -> Saved 'top_rated_movies_report.csv'.\n")
  
  # C. Popularity vs Rating
  cat("    -> Plotting: Popularity vs Average Rating...\n")
  p2 <- ggplot(movie_stats, aes(x = rating_count, y = average_rating)) +
    geom_point(alpha = 0.5, color = 'coral') +
    labs(title = 'Popularity vs Average Rating', x = 'Number of Ratings (Popularity)', y = 'Average Rating') +
    theme_minimal()
  ggsave('plots/2_popularity_vs_rating.png', plot = p2, width = 8, height = 5)
  
  # D. Genre Analysis
  cat("    -> Plotting: Genre Analysis...\n")
  # Splitting the pipe separated genre strings efficiently
  movie_genres <- movie_data %>%
    separate_rows(genres, sep = "\\|")
  
  genre_stats <- movie_genres %>%
    group_by(genres) %>%
    summarise(
      rating_count = n(),
      average_rating = mean(rating)
    ) %>%
    filter(rating_count > 1000) %>%
    arrange(desc(average_rating))
  
  # Ensure factors are sorted for plotting
  genre_stats$genres <- factor(genre_stats$genres, levels = genre_stats$genres)
  
  p3 <- ggplot(genre_stats, aes(x = genres, y = average_rating, fill = genres)) +
    geom_bar(stat = "identity") +
    scale_fill_viridis_d(option = "D") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position = "none") +
    labs(title = 'Average Rating by Genre (Min 1000 Ratings)', x = '', y = 'Average Rating')
  ggsave('plots/3_genre_analysis.png', plot = p3, width = 12, height = 6)
  
  # E. User Behavior Analysis
  cat("    -> Plotting: User Behavior Analysis...\n")
  user_stats <- movie_data %>%
    group_by(userId) %>%
    summarise(
      rating_count = n(),
      average_rating = mean(rating)
    )
  
  p4 <- ggplot(user_stats, aes(x = average_rating)) +
    geom_histogram(bins = 20, fill = 'purple', color = 'black', alpha = 0.8) +
    labs(title = 'Distribution of User Average Ratings', x = 'Average Rating given by User', y = 'Count of Users') +
    theme_minimal()
  ggsave('plots/4_user_behavior.png', plot = p4, width = 8, height = 5)
  
  cat("\n========================================\n")
  cat("          ANALYSIS COMPLETELY FINISHED        \n")
  cat("========================================\n")
  cat("All visualizations have been cleanly saved in the 'plots/' directory.\n")
  cat("Ready for your presentation/review!\n")
}

if (sys.nframe() == 0L) {
  main()
}
